File: classes/CoreLocationController.m
//
//  CoreLocationController.m
//  marguerite
//
//  Created by Kevin Conley on 6/26/13.
//  Copyright (c) 2013 Cardinal Devs. All rights reserved.
//

#import "CoreLocationController.h"


@implementation CoreLocationController
@synthesize locMgr, delegate;

- (id)init {
	self = [super init];
    
	if(self != nil) {
		self.locMgr = [[CLLocationManager alloc] init]; // Create new instance of locMgr
		self.locMgr.delegate = self; // Set the delegate as self.
	}
    
	return self;
}

- (void)locationManager:(CLLocationManager *)manager didUpdateToLocation:(CLLocation *)newLocation fromLocation:(CLLocation *)oldLocation {
	if([self.delegate conformsToProtocol:@protocol(CoreLocationControllerDelegate)]) {
        // Check if the class assigning itself as the delegate conforms to our protocol.  If not, the message will go nowhere.  Not good.
		[self.delegate locationUpdate:newLocation];
	}
}

- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error {
	if([self.delegate conformsToProtocol:@protocol(CoreLocationControllerDelegate)]) {
        // Check if the class assigning itself as the delegate conforms to our protocol.  If not, the message will go nowhere.  Not good.
		[self.delegate locationError:error];
	}
}

@end

File: Marguerite/CoreLocationController.swift
//
//  CoreLocationController.swift
//  A convenience class for receiving location updates via CLLocationManager.
//
//  Created by Kevin Conley on 3/2/15.
//  Copyright (c) 2015 Kevin Conley. All rights reserved.
//

import UIKit
import CoreLocation

protocol CoreLocationControllerDelegate: class {
    func locationUpdate(location: CLLocation)
    func locationError(error: NSError)
    func locationAuthorizationStatusChanged(nowEnabled: Bool)
}

class CoreLocationController: NSObject, CLLocationManagerDelegate {
    private let locationManager = CLLocationManager()
    
    weak var delegate: CoreLocationControllerDelegate?
    
    override init() {
        super.init()
        locationManager.delegate = self
        locationManager.startUpdatingLocation()
    }
    
    func refreshLocation() {
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        
        if CLLocationManager.authorizationStatus() == .NotDetermined {
            locationManager.requestWhenInUseAuthorization()
        }
        
        if CLLocationManager.locationServicesEnabled() {
            locationManager.startUpdatingLocation()
        }
    }

    func locationEnabled() -> Bool {
        return CLLocationManager.authorizationStatus() == .AuthorizedWhenInUse
    }
    
    func locationManager(manager: CLLocationManager!, didUpdateToLocation newLocation: CLLocation!, fromLocation oldLocation: CLLocation!) {
        delegate?.locationUpdate(newLocation)
        locationManager.stopUpdatingLocation()
    }
    
    func locationManager(manager: CLLocationManager!, didFailWithError error: NSError!) {
        delegate?.locationError(error)
        locationManager.stopUpdatingLocation()
    }
    
    func locationManager(manager: CLLocationManager!,
        didChangeAuthorizationStatus status: CLAuthorizationStatus) {
        if status == .AuthorizedWhenInUse {
            delegate?.locationAuthorizationStatusChanged(true)
        } else {
            delegate?.locationAuthorizationStatusChanged(false)
        }
    }
}
